core/editor/src/dispatcher/mod.rs

```diff
@@ -9,83 +9,74 @@ pub struct Dispatcher {
 }
 
 impl Dispatcher {
-	pub fn handle_event(&self, state: &mut EditorState, event: Event) -> Result<(), EditorError> {
+	pub fn handle_event(&self, tool_state: &mut ToolState, document: &mut Document, event: &Event) -> Result<(), EditorError> {
 		log::trace!("{:?}", event);
 
 		match event {
 			Event::SelectTool(tool_type) => {
-				state.tool_state.active_tool_type = tool_type;
-
-				Ok(())
+				tool_state.active_tool_type = *tool_type;
 			}
 			Event::SelectPrimaryColor(color) => {
-				state.tool_state.primary_color = color;
-
-				Ok(())
+				tool_state.primary_color = *color;
 			}
 			Event::SelectSecondaryColor(color) => {
-				state.tool_state.secondary_color = color;
-
-				Ok(())
+				tool_state.secondary_color = *color;
 			}
 			Event::SwapColors => {
-				std::mem::swap(&mut state.tool_state.primary_color, &mut state.tool_state.secondary_color);
-
-				Ok(())
+				std::mem::swap(&mut tool_state.primary_color, &mut tool_state.secondary_color);
 			}
 			Event::ResetColors => {
-				state.tool_state.primary_color = Color::BLACK;
-				state.tool_state.secondary_color = Color::WHITE;
-
-				Ok(())
+				tool_state.primary_color = Color::BLACK;
+				tool_state.secondary_color = Color::WHITE;
 			}
 			Event::MouseDown(mouse_state) => {
-				state.tool_state.mouse_state = mouse_state;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mouse_state = *mouse_state;
 			}
 			Event::MouseUp(mouse_state) => {
-				state.tool_state.mouse_state = mouse_state;
-
-				state.document.svg.push(SvgElement::Circle(Circle {
-					center: Point {
-						x: mouse_state.position.x as f64,
-						y: mouse_state.position.y as f64,
-					},
-					radius: 10.0,
-				}));
-				self.emit_response(Response::UpdateCanvas { document: state.document.render() });
-
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mouse_state = *mouse_state;
 			}
 			Event::MouseMovement(pos) => {
-				state.tool_state.mouse_state.position = pos;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mouse_state.position = *pos;
 			}
 			Event::ModifierKeyDown(mod_keys) => {
-				state.tool_state.mod_keys = mod_keys;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mod_keys = *mod_keys;
 			}
 			Event::ModifierKeyUp(mod_keys) => {
-				state.tool_state.mod_keys = mod_keys;
-				state.tool_state.active_tool()?.handle_input(event);
-
-				Ok(())
+				tool_state.mod_keys = *mod_keys;
 			}
 			Event::KeyPress(key) => todo!(),
```

Comment:
Let's turn this into `KeyDown` and `KeyUp`. I'm also thinking we should get rid of explicit `ModifierKeyDown`/`Up` events but still allow them to be queryable as the current input state. We can address this in a followup PR though.